File: exercises/03_unknown_repository/scripts/filtering_finalfinal_jolan.py
def filter_high_quality_variants(variants, min_depth=5):
    """
    Filters variant calls based on minimum read depth.
    There is a bug in this function where it modifies a dictionary while iterating over it.
    """
    for position in variants.keys():
        observations = variants[position]
        if len(observations) < min_depth:
            del variants[position]
            
    return variants

def write_vcf(filtered_variants, output_file):
    with open(output_file, 'w') as f:
        f.write("##fileformat=VCFv4.2\n")
        f.write("#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n")
        
        for pos, obs in filtered_variants.items():
            ref = obs[0]['ref']
            alt = obs[0]['alt']
            depth = len(obs)
            f.write(f"chr1\t{pos}\t.\t{ref}\t{alt}\t.\tPASS\tDP={depth}\n")

File: exercises/03_unknown_repository_R/scripts/filtering_finalfinal_jolan.R
filter_high_quality_variants <- function(variants, min_depth = 5) {
  # Filters variant calls based on minimum read depth.
  for (i in 1:length(variants)) {
    observations <- variants[[i]]
    if (length(observations) < min_depth) {
      variants[[i]] <- NULL
    }
  }
  return(variants)
}

write_vcf <- function(filtered_variants, output_file) {
  con <- file(output_file, "w")
  writeLines("##fileformat=VCFv4.2", con)
  writeLines("#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO", con)
  
  for (pos in names(filtered_variants)) {
    obs <- filtered_variants[[pos]]
    ref <- obs[[1]]$ref
    alt <- obs[[1]]$alt
    depth <- length(obs)
    writeLines(sprintf("chr1\t%s\t.\t%s\t%s\t.\tPASS\tDP=%d", pos, ref, alt, depth), con)
  }
  close(con)
}
